File: langs/c/src/tiss_hash.c
/*
 * tiss_hash.c - Implementacao do port C do hash MD5 do epilogo TISS/ANS.
 *
 * -----------------------------------------------------------------------------
 * Decisao de parser: libxml2
 * -----------------------------------------------------------------------------
 * Avaliadas tres opcoes:
 *
 * - libxml2 (ESCOLHIDA): parser de fato em Linux/Unix, mantido pela Gnome,
 *   usado por baixo do lxml (Python) e DOMDocument (PHP). Suporta:
 *     - DOM completo com xmlNodeGetContent normalizando entidades.
 *     - Comentarios preservados na arvore (xmlNode com type XML_COMMENT_NODE).
 *     - Resolucao correta de namespace W3C (ns->href).
 *     - Hardening contra XXE via flags (XML_PARSE_NONET sem NOENT-EXT).
 *     - Normalizacao automatica do encoding declarado pra UTF-8 internamente.
 *
 * - expat: SAX-based, exige reconstruir manualmente o conceito de folha
 *   (tracking de pilha + lookahead). Mais leve, mas zero ganho real
 *   pra XMLs TISS (< 5 MB tipico). Descartado.
 *
 * - yxml/mxml: parsers minimos sem suporte de namespace W3C correto.
 *   Descartados.
 *
 * - libxml++ (C++): wrap C++ — fora de escopo (este e port C puro).
 *
 * -----------------------------------------------------------------------------
 * Decisao de MD5: OpenSSL EVP API (libcrypto)
 * -----------------------------------------------------------------------------
 * - OpenSSL EVP_* (ESCOLHIDA): API moderna (a low-level MD5_* foi deprecada
 *   em OpenSSL 3.0). Disponivel em qualquer Linux, mantido, FIPS-ready.
 *
 * - Bundle RFC 1321 reference impl (~150 LOC, dominio publico): viavel pra
 *   eliminar dep de libcrypto. Mantido como opcao futura mas nao escolhido
 *   pra simplicidade — quem ja usa libxml2 ja tem libssl/libcrypto via
 *   ecossistema. Caller que quiser zero-libcrypto pode trocar este arquivo.
 *
 * -----------------------------------------------------------------------------
 * Buffer de concatenacao
 * -----------------------------------------------------------------------------
 * Implementado como struct `concat_buf_t` com crescimento por dobramento
 * (amortizado O(1) por append). Preferido sobre xmlBuffer da libxml2 porque:
 *   - controle direto de erro de alocacao (TISS_HASH_ERR_ALLOC).
 *   - sem dep transitiva de utf-8 specifics da libxml2.
 *   - profile-friendly: 1 realloc por crescimento, sem overhead de
 *     conversao interna.
 *
 * -----------------------------------------------------------------------------
 * Inicializacao thread-safe
 * -----------------------------------------------------------------------------
 * xmlInitParser() e idempotente porem nao thread-safe na PRIMEIRA chamada.
 * Usamos pthread_once_t (POSIX) pra garantir uma unica inicializacao.
 *
 * Licenca: MIT.
 */

#include "tiss_hash.h"

#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <libxml/parser.h>
#include <libxml/tree.h>
#include <libxml/xmlerror.h>

/* Suprime aviso de header da OpenSSL 3.x em algumas distros (FIPS macros). */
#define OPENSSL_NO_DEPRECATED 1
#include <openssl/evp.h>

/* -------------------------------------------------------------------------- */
/* Constantes internas                                                        */
/* -------------------------------------------------------------------------- */

#define TISS_HASH_VERSION_STR "0.1.0"

/* MD5 digest = 16 bytes. */
#define MD5_DIGEST_LEN 16

/* Capacidade inicial do buffer de concat (8 KiB). XMLs TISS tipicos
 * geram concats entre 1 KiB e 200 KiB; 8K cobre o caso comum sem realloc. */
#define CONCAT_BUF_INIT_CAP 8192

/* -------------------------------------------------------------------------- */
/* Inicializacao unica do parser libxml2                                       */
/* -------------------------------------------------------------------------- */

static pthread_once_t g_init_once = PTHREAD_ONCE_INIT;

static void tiss_hash_init_libxml(void)
{
    /* Inicializa estruturas globais do libxml2 (catalogos, hashes, etc).
     * Idempotente em chamadas subsequentes, mas a PRIMEIRA chamada nao
     * e thread-safe — por isso o once. */
    xmlInitParser();

    /* Silenciar handler de erro default (escreve em stderr). Vamos
     * relatar erro via codigo de retorno; mensagens do parser ficam
     * acessiveis via xmlGetLastError() se o caller quiser instrumentar. */
    xmlSetGenericErrorFunc(NULL, NULL);
}

/* -------------------------------------------------------------------------- */
/* Buffer dinamico de concatenacao                                            */
/* -------------------------------------------------------------------------- */

typedef struct {
    char  *data;
    size_t len;
    size_t cap;
    int    oom; /* 1 se uma alocacao falhou; 0 caso contrario */
} concat_buf_t;

static void concat_init(concat_buf_t *b)
{
    b->data = (char *)malloc(CONCAT_BUF_INIT_CAP);
    b->len  = 0;
    b->cap  = b->data ? CONCAT_BUF_INIT_CAP : 0;
    b->oom  = b->data ? 0 : 1;
}

static void concat_free(concat_buf_t *b)
{
    free(b->data);
    b->data = NULL;
    b->len  = 0;
    b->cap  = 0;
}

/* Garante capacidade pra escrever mais `extra` bytes. */
static void concat_ensure(concat_buf_t *b, size_t extra)
{
    if (b->oom) {
        return;
    }
    /* Overflow check: len + extra nao pode estourar size_t. */
    if (extra > SIZE_MAX - b->len) {
        b->oom = 1;
        return;
    }
    size_t need = b->len + extra;
    if (need <= b->cap) {
        return;
    }
    size_t new_cap = b->cap ? b->cap : CONCAT_BUF_INIT_CAP;
    while (new_cap < need) {
        if (new_cap > SIZE_MAX / 2) {
            b->oom = 1;
            return;
        }
        new_cap *= 2;
    }
    char *new_data = (char *)realloc(b->data, new_cap);
    if (!new_data) {
        b->oom = 1;
        return;
    }
    b->data = new_data;
    b->cap  = new_cap;
}

static void concat_append(concat_buf_t *b, const char *s, size_t n)
{
    if (n == 0) {
        return;
    }
    concat_ensure(b, n);
    if (b->oom) {
        return;
    }
    memcpy(b->data + b->len, s, n);
    b->len += n;
}

/* -------------------------------------------------------------------------- */
/* Logica de folha + walker                                                   */
/* -------------------------------------------------------------------------- */

/* Verifica se um filho conta como "estruturado" pra desclassificar o pai
 * como folha. Espelha a logica do PHP/Rust ports:
 *   - XML_ELEMENT_NODE, XML_COMMENT_NODE, XML_PI_NODE contam.
 *   - Text/CDATA NAO contam (TISS sem mixed content; um elemento com so
 *     text dentro e folha de valor).
 */
static int is_structured_child(const xmlNode *child)
{
    switch (child->type) {
    case XML_ELEMENT_NODE:
    case XML_COMMENT_NODE:
    case XML_PI_NODE:
        return 1;
    default:
        return 0;
    }
}

/* Retorna 1 se o no e "folha pro hash": Element OU Comment sem filhos
 * estruturados. */
static int is_leaf_for_hash(const xmlNode *n)
{
    if (n->type != XML_ELEMENT_NODE && n->type != XML_COMMENT_NODE) {
        return 0;
    }
    for (const xmlNode *c = n->children; c != NULL; c = c->next) {
        if (is_structured_child(c)) {
            return 0;
        }
    }
    return 1;
}

/* Verifica se este elemento e o <ans:hash> do namespace TISS. */
static int is_tiss_hash_element(const xmlNode *n)
{
    if (n->type != XML_ELEMENT_NODE) {
        return 0;
    }
    if (n->name == NULL || strcmp((const char *)n->name, "hash") != 0) {
        return 0;
    }
    if (n->ns == NULL || n->ns->href == NULL) {
        return 0;
    }
    return strcmp((const char *)n->ns->href, TISS_HASH_NAMESPACE) == 0;
}

/* Acha o primeiro <ans:hash> em ordem de documento (DFS pre-order). */
static xmlNode *find_first_tiss_hash(xmlNode *node)
{
    for (xmlNode *n = node; n != NULL; n = n->next) {
        if (is_tiss_hash_element(n)) {
            return n;
        }
        if (n->children) {
            xmlNode *found = find_first_tiss_hash(n->children);
            if (found) {
                return found;
            }
        }
    }
    return NULL;
}

/* Remove todos os filhos de um nodo (usado pra "zerar" <ans:hash>). */
static void clear_node_children(xmlNode *n)
{
    xmlNode *c = n->children;
    while (c != NULL) {
        xmlNode *next = c->next;
        xmlUnlinkNode(c);
        xmlFreeNode(c);
        c = next;
    }
}

/* Apend do texto da folha ao buffer.
 *
 * - Elemento-folha: usa xmlNodeGetContent (concatena Text/CDATA filhos,
 *   ja decodificados em UTF-8). Para folha sem filhos, retorna "".
 *   Para folha com so Text/CDATA dentro, retorna o texto literal.
 * - Comentario-folha: n->content e o texto entre <!-- e --> (UTF-8).
 *
 * libxml2 normaliza CR/LF (XML 1.0 §2.11) durante o parse. CR/CRLF viram
 * LF antes de chegar aqui — comportamento alinhado com a referencia (que
 * tambem confia na normalizacao do parser). */
static void append_leaf_text(concat_buf_t *buf, const xmlNode *leaf)
{
    if (leaf->type == XML_COMMENT_NODE) {
        if (leaf->content) {
            concat_append(buf, (const char *)leaf->content,
                          strlen((const char *)leaf->content));
        }
        return;
    }

    /* Elemento-folha: xmlNodeGetContent aloca uma copia do texto. */
    xmlChar *txt = xmlNodeGetContent(leaf);
    if (txt) {
        concat_append(buf, (const char *)txt, strlen((const char *)txt));
        xmlFree(txt);
    }
    /* Se xmlNodeGetContent retornar NULL, equivale a "" — nao apendamos
     * nada (string vazia + concat = concat). */
}

/* Walker recursivo: em ordem de documento, encontra folhas (Element/Comment
 * sem filhos estruturados) e apenda seu texto. Pula o no <ans:hash> (cujo
 * conteudo ja foi zerado externamente, mas o pulo aqui e defensivo: ainda
 * que vazio, mantemos a semantica "<ans:hash> contribui string vazia").
 *
 * Nota: como zeramos os filhos do <ans:hash> via clear_node_children, ele
 * vira folha (sem children) e o append_leaf_text apendaria "" via
 * xmlNodeGetContent. O resultado e o mesmo, mas o skip explicito documenta
 * a intencao e protege contra futuras mudancas. */
static void walk(xmlNode *node, concat_buf_t *buf, const xmlNode *hash_node)
{
    for (xmlNode *n = node; n != NULL; n = n->next) {
        if (n->type != XML_ELEMENT_NODE && n->type != XML_COMMENT_NODE) {
            /* Text, CDATA, PI no topo nao sao iterados como folhas (eles
             * aparecem so como CONTEUDO de folhas Element via
             * xmlNodeGetContent). DTD/Entity nodes etc. tambem sao
             * pulados. */
            continue;
        }
        if (is_leaf_for_hash(n)) {
            if (n == hash_node) {
                /* <ans:hash> contribui string vazia (ja zerado). */
                continue;
            }
            append_leaf_text(buf, n);
        } else if (n->children) {
            /* Nao-folha: descer recursivo. */
            walk(n->children, buf, hash_node);
        }
    }
}

/* -------------------------------------------------------------------------- */
/* MD5 (OpenSSL EVP) + hex encoding                                           */
/* -------------------------------------------------------------------------- */

/* Converte 16 bytes do digest em 32 hex lowercase + NUL. */
static void hex_lower_encode(const unsigned char *digest, char out[TISS_HASH_HEX_LEN])
{
    static const char HEX[] = "0123456789abcdef";
    for (int i = 0; i < MD5_DIGEST_LEN; i++) {
        out[i * 2]     = HEX[(digest[i] >> 4) & 0x0F];
        out[i * 2 + 1] = HEX[ digest[i]       & 0x0F];
    }
    out[MD5_DIGEST_LEN * 2] = '\0';
}

/* Calcula MD5 dos bytes em (data, len) e escreve hex lowercase em out_hex.
 * Retorna 0 em sucesso, -1 em falha (alocacao do EVP_MD_CTX). */
static int md5_hex(const unsigned char *data, size_t len,
                   char out_hex[TISS_HASH_HEX_LEN])
{
    EVP_MD_CTX *ctx = EVP_MD_CTX_new();
    if (!ctx) {
        return -1;
    }

    unsigned char raw[MD5_DIGEST_LEN];
    unsigned int  raw_len = 0;
    int           ok      = 0;

    /* impl=NULL escolhe a impl default (libcrypto default ENGINE). */
    if (EVP_DigestInit_ex(ctx, EVP_md5(), NULL) != 1) {
        goto out;
    }
    if (len > 0 && EVP_DigestUpdate(ctx, data, len) != 1) {
        goto out;
    }
    if (EVP_DigestFinal_ex(ctx, raw, &raw_len) != 1) {
        goto out;
    }
    if (raw_len != MD5_DIGEST_LEN) {
        goto out;
    }
    ok = 1;

out:
    EVP_MD_CTX_free(ctx);
    if (!ok) {
        return -1;
    }
    hex_lower_encode(raw, out_hex);
    return 0;
}

/* -------------------------------------------------------------------------- */
/* API publica                                                                */
/* -------------------------------------------------------------------------- */

tiss_hash_status_t tiss_hash_bytes(const unsigned char *xml,
                                   size_t len,
                                   char out_hex[TISS_HASH_HEX_LEN])
{
    if (xml == NULL || out_hex == NULL || len == 0) {
        return TISS_HASH_ERR_ARG;
    }
    /* libxml2 usa `int` pra tamanho em xmlReadMemory; rejeitar entrada
     * que estoura. Limite pratico TISS << INT_MAX. */
    if (len > (size_t)0x7FFFFFFF) {
        return TISS_HASH_ERR_ARG;
    }

    pthread_once(&g_init_once, tiss_hash_init_libxml);

    /* Hardening:
     *   XML_PARSE_NONET  - proibe rede (file://, http://) em resolucao
     *                      de DTD/entidade.
     *   XML_PARSE_NOENT  - substitui entidades predefinidas (&amp; etc.)
     *                      pelo valor — necessario pra reproduzir
     *                      ambiguidade #4 (entidades sao decodificadas).
     *
     * NAO usamos XML_PARSE_NOBLANKS: perderia whitespace dentro de
     * valores (ambiguidade #7).
     *
     * NAO usamos XML_PARSE_RECOVER em producao: queremos falhar em XML
     * mal-formado. */
    const int parse_flags = XML_PARSE_NONET | XML_PARSE_NOENT;

    xmlResetLastError();

    /* url=NULL, encoding=NULL deixa libxml2 detectar via declaracao XML
     * (geralmente iso-8859-1). O parser converte tudo pra UTF-8 internamente,
     * entao xmlNodeGetContent ja devolve UTF-8 — exatamente o que precisamos
     * pro MD5. */
    xmlDoc *doc = xmlReadMemory((const char *)xml, (int)len,
                                NULL, NULL, parse_flags);
    if (!doc) {
        return TISS_HASH_ERR_INVALID_XML;
    }

    xmlNode *root = xmlDocGetRootElement(doc);
    if (!root) {
        xmlFreeDoc(doc);
        return TISS_HASH_ERR_INVALID_XML;
    }

    /* 1) Zerar conteudo do primeiro <ans:hash>. Multiplos <ans:hash>
     *    no documento = comportamento NAO fixado (ambiguidade #9);
     *    seguimos a referencia e zeramos so o primeiro. */
    xmlNode *hash_node = find_first_tiss_hash(root);
    if (hash_node) {
        clear_node_children(hash_node);
    }

    /* 2) Walker em ordem de documento. */
    concat_buf_t buf;
    concat_init(&buf);
    if (buf.oom) {
        xmlFreeDoc(doc);
        return TISS_HASH_ERR_ALLOC;
    }

    walk(root, &buf, hash_node);

    tiss_hash_status_t rc = TISS_HASH_OK;
    if (buf.oom) {
        rc = TISS_HASH_ERR_ALLOC;
        goto cleanup;
    }

    /* 3) MD5 dos bytes UTF-8 (libxml2 ja entregou em UTF-8). */
    if (md5_hex((const unsigned char *)buf.data, buf.len, out_hex) != 0) {
        rc = TISS_HASH_ERR_ALLOC;
        goto cleanup;
    }

cleanup:
    concat_free(&buf);
    xmlFreeDoc(doc);
    return rc;
}

tiss_hash_status_t tiss_hash_file(const char *path,
                                  char out_hex[TISS_HASH_HEX_LEN])
{
    if (path == NULL || out_hex == NULL) {
        return TISS_HASH_ERR_ARG;
    }

    FILE *fp = fopen(path, "rb");
    if (!fp) {
        return TISS_HASH_ERR_IO;
    }

    if (fseek(fp, 0, SEEK_END) != 0) {
        fclose(fp);
        return TISS_HASH_ERR_IO;
    }
    long size_l = ftell(fp);
    if (size_l < 0) {
        fclose(fp);
        return TISS_HASH_ERR_IO;
    }
    if (fseek(fp, 0, SEEK_SET) != 0) {
        fclose(fp);
        return TISS_HASH_ERR_IO;
    }

    if (size_l == 0) {
        fclose(fp);
        return TISS_HASH_ERR_INVALID_XML;
    }

    size_t         size = (size_t)size_l;
    unsigned char *buf  = (unsigned char *)malloc(size);
    if (!buf) {
        fclose(fp);
        return TISS_HASH_ERR_ALLOC;
    }

    size_t n_read = fread(buf, 1, size, fp);
    int    rerr   = ferror(fp);
    fclose(fp);

    if (rerr || n_read != size) {
        free(buf);
        return TISS_HASH_ERR_IO;
    }

    tiss_hash_status_t rc = tiss_hash_bytes(buf, size, out_hex);
    free(buf);
    return rc;
}

const char *tiss_hash_strerror(tiss_hash_status_t s)
{
    switch (s) {
    case TISS_HASH_OK:              return "ok";
    case TISS_HASH_ERR_INVALID_XML: return "XML invalido ou mal-formado";
    case TISS_HASH_ERR_IO:          return "falha de I/O ao ler arquivo";
    case TISS_HASH_ERR_ALLOC:       return "falha de alocacao de memoria";
    case TISS_HASH_ERR_ARG:         return "argumento invalido";
    default:                        return "erro desconhecido";
    }
}

const char *tiss_hash_version(void)
{
    return TISS_HASH_VERSION_STR;
}

File: langs/go/tisshash.go
// Package tisshash: implementacao Go do hash MD5 do epilogo TISS/ANS.
//
// Ver doc.go para a descricao geral do package e o algoritmo canonico.
package tisshash

import (
	"bytes"
	"crypto/md5"
	"encoding/hex"
	"encoding/xml"
	"errors"
	"fmt"
	"io"
	"os"
	"strings"

	"golang.org/x/text/encoding/charmap"
)

// Namespace XML do Padrao TISS/ANS. Usado para localizar <ans:hash>
// independente do prefixo escolhido pelo emissor (o que importa eh o URI).
const Namespace = "http://www.ans.gov.br/padroes/tiss/schemas"

// Version desta implementacao.
const Version = "0.1.0"

// InvalidTissXMLError sinaliza que o XML eh mal-formado ou nao pode ser
// decodificado/parseado para calcular o hash. Preserva a causa raiz via
// Unwrap() para inspecao por errors.Is/As.
type InvalidTissXMLError struct {
	Err error
}

// Error implementa error.
func (e *InvalidTissXMLError) Error() string {
	if e == nil || e.Err == nil {
		return "tisshash: XML invalido"
	}
	return "tisshash: XML invalido: " + e.Err.Error()
}

// Unwrap permite errors.Is/As atravessar para a causa raiz do parser.
func (e *InvalidTissXMLError) Unwrap() error {
	if e == nil {
		return nil
	}
	return e.Err
}

// HashTiss calcula o hash MD5 canonico do epilogo TISS/ANS a partir dos bytes
// do XML completo. Retorna 32 chars hex lowercase.
//
// Decisoes / alternativas avaliadas:
//
//   - Parser: encoding/xml da stdlib em modo streaming (Decoder.Token()).
//     Zero dependencia externa. Tokens emitidos cobrem todos os tipos de no
//     necessarios pra reproduzir a semantica da referencia Python: StartElement,
//     EndElement, CharData (texto + CDATA), Comment, ProcInst, Directive.
//     Alternativas descartadas:
//
//   - github.com/beevik/etree: DOM 3rd-party, ergonomico mas dep externa
//     desnecessaria pra este escopo.
//
//   - github.com/antchfx/xmlquery: focado em XPath, overkill.
//
//   - golang.org/x/net/html: parser HTML, nao XML.
//
//   - MD5: crypto/md5 da stdlib (zero dep).
//
//   - Encoding ISO-8859-1: a stdlib so aceita UTF-8 por default; instalamos
//     um Decoder.CharsetReader que delega pra charmap.ISO8859_1 do
//     golang.org/x/text. Esta eh a forma idiomatica em Go.
//
// Ver conformance/AMBIGUITY_NOTES.md no repositorio para o catalogo das 15
// decisoes canonicas (comentario entra, atributo nao entra, CDATA literal,
// prefixo de namespace irrelevante, bytes UTF-8 nao ISO-8859-1, etc.).
func HashTiss(xmlBytes []byte) (string, error) {
	// Strip BOM UTF-8 (TISS proibe, mas a referencia aceita).
	raw := stripBOM(xmlBytes)

	dec := xml.NewDecoder(bytes.NewReader(raw))
	dec.CharsetReader = charsetReader
	// Politica de seguranca: nao expandir entidades externas / DTDs externos.
	// encoding/xml ja eh restrito por design (nao busca DTDs remotos), mas
	// reforcamos desativando autoclose e mantendo entidades padrao apenas.
	dec.Strict = true

	// Estado de parse: pilha de elementos abertos. Cada elemState acumula
	// texto e marca se viu filho estruturado (Element/Comment/PI/Directive),
	// que eh o criterio pra desclassificar como "folha".
	//
	// Invariante: o topo da pilha eh sempre o elemento corrente em cujo
	// escopo os proximos tokens devem ser processados. Pop ocorre em
	// EndElement.
	type elemState struct {
		name           xml.Name
		isAnsHash      bool
		hasStructChild bool
		text           strings.Builder
	}

	var (
		stack     []*elemState
		concat    strings.Builder
		foundHash bool // ja zeramos o primeiro <ans:hash>? (politica: so o primeiro)
	)

	// markParentStructured: quando ha um filho do tipo Element/Comment/PI/
	// Directive, o pai deixa de ser folha. Replica o filtro `len(el) == 0`
	// do lxml, onde len() conta children Element/Comment/PI.
	markParentStructured := func() {
		if n := len(stack); n > 0 {
			stack[n-1].hasStructChild = true
		}
	}

	for {
		tok, err := dec.Token()
		if err == io.EOF {
			break
		}
		if err != nil {
			return "", &InvalidTissXMLError{Err: err}
		}

		switch t := tok.(type) {
		case xml.StartElement:
			// Antes de empilhar o novo elemento, marca pai como estruturado.
			markParentStructured()
			es := &elemState{
				name: t.Name,
				isAnsHash: t.Name.Space == Namespace &&
					t.Name.Local == "hash" &&
					!foundHash,
			}
			if es.isAnsHash {
				foundHash = true
			}
			stack = append(stack, es)

		case xml.EndElement:
			n := len(stack)
			if n == 0 {
				// Robustez: nao deveria acontecer com XML bem-formado, mas
				// o parser ja teria errado antes de chegar aqui.
				return "", &InvalidTissXMLError{
					Err: fmt.Errorf("EndElement sem StartElement correspondente: %s", t.Name.Local),
				}
			}
			top := stack[n-1]
			stack = stack[:n-1]

			// Eh folha se nao teve nenhum filho estruturado.
			if !top.hasStructChild {
				if top.isAnsHash {
					// Zerar conteudo do <ans:hash>: contribui "" (no-op
					// efetivo, mas explicito pra leitura do algoritmo).
					concat.WriteString("")
				} else {
					concat.WriteString(top.text.String())
				}
			}

		case xml.CharData:
			// Texto e CDATA chegam ambos como CharData. Acumula no topo da
			// pilha SOMENTE; sera contribuido pro concat global apenas se o
			// elemento terminar sendo classificado como folha (caso contrario
			// eh whitespace de indentacao entre tags filhas, que NAO entra).
			if n := len(stack); n > 0 {
				stack[n-1].text.Write(t)
			}
			// Se estiver fora de qualquer elemento (entre prologo e root, ou
			// pos-root), ignoramos: eh whitespace estrutural.

		case xml.Comment:
			// Comentario eh "folha de comentario": contribui seu texto direto
			// ao concat global E marca o pai como estruturado (desclassifica
			// o pai como folha, replicando lxml onde Comment conta em len()).
			//
			// Esta eh a ambiguidade #2 fixada pela referencia: comentarios
			// XML ENTRAM no concat. Ver AMBIGUITY_NOTES.md.
			markParentStructured()
			concat.Write(t)

		case xml.ProcInst:
			// Processing instruction (ex.: <?php ... ?>). Marca o pai como
			// estruturado (desclassifica como folha). Nao contribuimos texto
			// (seguindo o port Rust, que trata so Element/Comment como
			// nos-folha que contribuem). Nenhum vetor de teste cobre PI no
			// corpo, entao este eh comportamento defensivo conservador.
			markParentStructured()

		case xml.Directive:
			// Diretiva DOCTYPE/etc. Marca pai como estruturado, nao
			// contribui. Tambem nao coberto por vetores.
			markParentStructured()
		}
	}

	// MD5 dos bytes UTF-8 da string concatenada.
	sum := md5.Sum([]byte(concat.String()))
	return hex.EncodeToString(sum[:]), nil
}

// HashTissFile eh atalho: le arquivo do disco e chama HashTiss.
//
// Retorna InvalidTissXMLError se o conteudo for invalido, ou erro de I/O
// (envelopado em fmt.Errorf com contexto do path) se a leitura falhar.
func HashTissFile(path string) (string, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return "", fmt.Errorf("tisshash: ler %q: %w", path, err)
	}
	return HashTiss(raw)
}

// -- Internos -------------------------------------------------------------

// stripBOM remove o BOM UTF-8 (EF BB BF) se presente no inicio dos bytes.
//
// TISS proibe BOM (encoding declarado eh ISO-8859-1, que nao tem BOM), mas
// a referencia aceita defensivamente.
func stripBOM(b []byte) []byte {
	if len(b) >= 3 && b[0] == 0xEF && b[1] == 0xBB && b[2] == 0xBF {
		return b[3:]
	}
	return b
}

// charsetReader eh o hook que encoding/xml usa quando ve uma declaracao
// encoding="..." diferente de UTF-8 no prologo.
//
// Suportamos:
//   - iso-8859-1 / latin1 (TISS canonico)
//   - utf-8 / us-ascii (passa direto)
//
// Outros encodings retornam erro explicito: melhor falhar cedo do que
// produzir hash silenciosamente errado.
func charsetReader(charset string, input io.Reader) (io.Reader, error) {
	cs := strings.ToLower(strings.TrimSpace(charset))
	switch cs {
	case "iso-8859-1", "latin1", "iso_8859-1", "iso-ir-100":
		return charmap.ISO8859_1.NewDecoder().Reader(input), nil
	case "utf-8", "utf8", "us-ascii", "ascii", "":
		return input, nil
	default:
		return nil, errors.New("tisshash: encoding nao suportado: " + charset)
	}
}
